### ISSUE TITLE
Question about the deferred shading influence

### ISSUE THREAD

**Issue by fzy28** [cgtuebingen/SSS-GS#2]
First thanks for the amazing work!

I am confused about why point-based rendering can not learn the specular while you mentioned the deferred shading would help.

Based on my understanding, though the roughness and albedo remains unchanged in the 3DGS, but we have the gaussian kernel function for help, and the original 3DGS was able to capture some very high frequency results.

I might miss something, though I can understand the intuition that 3DGS might be a large area and we might suffer from only using result around the center. 

I don't know how you implement the adaptive GS control (add and prune stuff).  If you use the original 3DGS version, then you don't take the gradient of roughness, albedo stuff into account, since the original 3DGS paper will generate additional points to simulate high frequency thing based on these gradient...

**Comment by JDihlmann** [comment]
Hey, thanks for reading our work and the interest! I'll try my best explaining it, but feel free to ask more questions. 

**What is the issue?**
So the default 3DGS pipeline only has the color represented as SHs allowing for view dependent effects. However, the choice of SHs already does not allow for high frequent details. Further if your 3D Gaussians are large that means that this view dependent effects will get blurred even more over the entire area of the Gaussians. Sure the original 3DGS can place very small Gaussians to mimic high frequent view dependent effect - **but** this is not possible for relightable methods were those effects (e.g. specular highlights) "travel over" the surface for different light settings. These are the reasons why Relightable 3D GS fails at representing very high details. 

**Why deferred shading? How does it solve the issue?**
If we evaluate the shader on the Gaussians and then rasterize high frequent details will get blurred over the entire Gaussian again, and therefore are soft in the final rendering. However if we rasterize and then apply the shader we are setting the final color directly for each pixel not the Gaussian, also having more "resolution" as we are not taking e.g. the view direction per Gaussian but per pixel. So the difference is that without this attentional blur we can have very crisp details in pixel space. 


![ablation](https://github.com/user-attachments/assets/573c49a9-6ac4-4039-b1cb-8328dcc73865)

Here is again our Ablation on this you could think of it as:
-  w/o PBR => 3DGS (+ relightable)
- w/o Deferred => R3DGS
- w/ Deferred => Ours

**Comment by fzy28** [comment]
Thanks for you kind reply!

But I still have not totally understood. Could you please explain a bit more about the description **"travel over" the surface for different light settings**, I don't quite get it, what do you mean by "travel over"?

And I feel like if you set the rasterized color and do deferred shading, here you treat the GS as a basis function like neural network to store and fit the roughness map etc, am I correct?

If I am correct, can we directly use a network to store these value since it is per pixel, dimension is quite low and the compression should be much higher, saving a lot of memory.

**Comment by JDihlmann** [comment]
Sorry for the high-level description. I thought it would be more easily understandable that way. By traveling over the surface I meant effects like specular highlights that move with the light direction (see attached video).

https://github.com/user-attachments/assets/f7bd1453-b00f-48f8-a269-81393227749d

I don't fully get your second question you might elaborate on this a bit more. Each Gaussian currently stores its base-color (3 channels) RGB, roughness (1), metalness (1), normal (3), subsurfaceness (1), visibility (SHs), roation & scale (5) we have a global MLP that receives some of those properties including the light & view direction and positional encoding of the gaussian as input. Returning the incident light and the SSS residual. Some initial properties and the MLP computed properties are rasterized and used in our shading model. See this alternative pipeline for a deeper understanding.
![pipeline](https://github.com/user-attachments/assets/ff0507b1-bf6c-49d3-857a-ed43cddf5ed7)

Sure we could use a network to store these value, however we wanted to be more comparable to the original GS method and aimed for high quality results and not further compression.

**Comment by fzy28** [comment]
Hi, thanks for you kind reply!

I think I understand the "travel over" meaning. But it makes me even more confused lol. Since based on this, the specularity will highly based on the the view direction (as well as the light direction), thus when you render a image from one view, it will have certain case that the specular part was quite wrong, causing large gradient, and the GS points should copy themselves to mimic the high frequency changing of the roughness and metallic.

I might have another thought about the reason, could it be the deferred shading you will obtain a better normal? Since the original 3DGS does not have specific definition about normal, and render on each point, whose normal might be quite bad, but the weighted sum normal is correct, and that's how we train it, using the weighted sum to supervise.

**Comment by JDihlmann** [comment]
There might be a significant gradient in one specific case, but this will be mitigated over multiple iterations with different view and light directions. On the contrary, we need to actively work to obtain a strong signal to aim for specularity (as you can also see in our metrics in the supplemental ablation, only showing minor differences compared to the visual ablation I posted above). Additionally, we freeze the roughness for the first 10k iterations to initially optimize geometry, and then focus on high-frequency appearance. However, I don’t see how your concern would specifically apply to deferred shading.

You’re absolutely correct that deferred shading also impacts the gradients of the normals, though we do not specifically claim that since other methods have already observed this (DeferredGS / 3D Gaussian splatting with deferred reflection).

Yes, the FPS is only claimed for OLAT, but it is still notably faster than previous NeRF-based methods. Additionally, you can pre-integrate the illumination and then apply the environment map with a simple matrix multiplication. That said, you’re totally right—there are exciting directions to explore, such as multiple lights, different intensities, and colors. We are also eager to look into multi-spectral decomposition / learning.

**Comment by fzy28** [comment]
I know its small but still if I understand it correctly, that on each Gaussian there are 3 MLPs, and the largest one with 3-layer and each layer with [64, 32, 32] neurons, in total more than 6000 floaters to store on each Gaussian?

So have many gaussian points for your each scene, I fear 24GB memory may not even enough for large scene...

I am not sure if I understand it correctly... And I am quite curious about how you make such case real-time doable,  really expecting your code released!

It would be the best if you can give a bit detailed implementation to show, are you using the tinymlp things? since I feel though mlp is small, but it is still quite large for real-time application based on my experience.

Thanks for your reply in advance!

**Comment by JDihlmann** [comment]
It's a global MLP with two different heads for all Gaussians not per Gaussian. Here is the implementation of the MLP, full clean code with follow soon (this is an uncleaned version):

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import math

class SSS(nn.Module):
    def __init__(self, input_size: int = 1, output_size: int = 1, net_width: int = 32): # TODO: Change the default values
        super().__init__()
        self.input_size = input_size
        self.output_size = output_size
        self.net_width = net_width
        self.factor = 2

        self.degree = 4

        
        self.main = nn.Sequential(
            nn.Linear(18 + 3 * 2 * self.degree, self.factor * self.net_width),
            nn.LeakyReLU(),
            nn.Linear(self.net_width * self.factor, self.net_width),
            nn.LeakyReLU(),
            nn.Linear(self.net_width, self.net_width),
            nn.LeakyReLU(),
        )

        self.residual = nn.Sequential(
            nn.Linear(self.net_width, 3),
            nn.Sigmoid(),
        )

        self.incident_light = nn.Sequential(
            nn.Linear(self.net_width, self.net_width),
            nn.LeakyReLU(),
            nn.Linear(self.net_width, 1),
            nn.ReLU()
        )

    def positional_encoding(self, x):
        """ Positional encoding of the inputs """
        result = []
        for d in range(self.degree):
            for fn in [torch.sin, torch.cos]:
                result.append(fn(2.0 ** d * math.pi * x))
        return torch.cat(result, dim=-1)


    def forward(self,  positions, rotations, scales, view_dirs, light_dirs, normals, visibilities, light_distances, iteration=-1):
        """ Forward pass of the model

        Args:
            positions: # [batch, 3]
            rotations: # [batch, 2]
            scales: # [batch, 3]
            view_dirs: # [batch, 3]
            light_dirs: # [batch, 3]
            normals: # [batch, 3]
            visibilities: # [batch, 1]
            light_distances: # [batch, 1]
        """

        encoded_positions = self.positional_encoding(positions)

        x = torch.concat([encoded_positions, rotations, scales, view_dirs, light_dirs, normals, visibilities, light_distances], dim=-1)
        x = self.main(x)

        residual = self.residual(x)
        incident_light = self.incident_light(x).repeat(1, 3)
 
        return residual, incident_light
        
```
